src/protocols/rdp/guac_rdpdr/rdpdr_messages.c

```diff
@@ -122,20 +124,28 @@ static void guac_rdpdr_send_client_capability(guac_rdpdrPlugin* rdpdr) {
 
 static void guac_rdpdr_send_client_device_list_announce_request(guac_rdpdrPlugin* rdpdr) {
 
-    int i;
-    wStream* output_stream = Stream_New(NULL, 256);
+    /* Calculate number of bytes needed for the stream */
+    int streamBytes = 16;
+    for (int i=0; i < rdpdr->devices_registered; i++)
+        streamBytes += rdpdr->devices[i].device_announce_len;
+
+    /* Allocate the stream */
+    wStream* output_stream = Stream_New(NULL, streamBytes);
 
     /* Write header */
     Stream_Write_UINT16(output_stream, RDPDR_CTYP_CORE);
     Stream_Write_UINT16(output_stream, PAKID_CORE_DEVICELIST_ANNOUNCE);
 
-    /* List devices */
+    /* Get the stream for each of the devices. */
     Stream_Write_UINT32(output_stream, rdpdr->devices_registered);
-    for (i=0; i<rdpdr->devices_registered; i++) {
-        guac_rdpdr_device* device = &(rdpdr->devices[i]);
-        device->announce_handler(device, output_stream, i);
+    for (int i=0; i<rdpdr->devices_registered; i++) {
+        
+        Stream_Copy(output_stream, rdpdr->devices[i].device_announce,
```

Comment:
I'm not sure `Stream_Copy()` is the right function/macro to use in this case. Checking the source for FreeRDP 1.1, I see `Stream_Copy()` defined as:

```c
#define Stream_Copy(_dst, _src, _n) do { \
    memcpy(_dst->pointer, _src->pointer, _n); \
    _dst->pointer += _n; \
    _src->pointer += _n; \
    } while (0)
```

It thus depends on the current write position within the buffer, and updates that write position as a result of the call.

Should we be using `Stream_Write()` and `Stream_Buffer()` instead?

Reply:
I'll take a look - I might need some guidance figuring this out, but I'll give it a try.

Reply:
Well, clearly I've messed something up - not sure if it's this or the printer naming thing, but segfaults every time...

Reply:
:shipit:

Reply:
I'll pull what you've got so far and see if something jumps out.

Reply:
...am reverting c30fd7a first, though. Writing the raw data structure is a bit too scary, especially since it only contains a pointer to the desired buffer. It shouldn't work, and I'm not sure what I'd do if it _did_ work. ;)

Reply:
aHA! You're missing the `Stream_New()` and corresponding assignment to `device->device_announce` for the printer device. It's there for filesystem, but not for printer:

https://github.com/apache/guacamole-server/blob/e2994778c3fed6e71f809aa589afe2db32b2e0ae/src/protocols/rdp/guac_rdpdr/rdpdr_fs_service.c#L144

Reply:
Quickly added the necessary allocation and assignment for the sake of testing, and it seems to work nicely:

![renamed-printer](https://user-images.githubusercontent.com/4632905/42199769-2f955844-7e45-11e8-86b2-ea5ba779d591.png)

Reply:
:man_facepalming: Doh.  Yeah, that usually helps.  Okay, I've initialized this and reverted my bad try below.  Should be good, now.